api/v1alpha1/deployment_types.go

```diff
@@ -0,0 +1,73 @@
+package v1alpha1
+
+import (
+	apps "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+const (
+	// DeploymentStrategyAnnotation is annotation for deployment,
+	// which is strategy fields of Advanced Deployment.
+	DeploymentStrategyAnnotation = "rollouts.kruise.io/deployment-strategy"
+
+	// DeploymentExtraStatusAnnotation is annotation for deployment,
+	// which is extra status field of Advanced Deployment.
+	DeploymentExtraStatusAnnotation = "rollouts.kruise.io/deployment-extra-status"
+
+	// DeploymentRolloutPolicyAnnotation must be "batch" or "canary":
+	// * "partitioned" means rolling in batches just like CloneSet, and will NOT create any extra Deployment;
+	// * "canary" means rolling in canary way, and will create a canary Deployment.
+	// Defaults to canary
+	DeploymentRolloutPolicyAnnotation = "rollouts.kruise.io/deployment-rolling-policy"
+)
+
+// DeploymentStrategy is strategy field for Advanced Deployment
+type DeploymentStrategy struct {
+	// original deployment strategy fields
+	apps.DeploymentStrategy
+	// Paused = true will block the upgrade of Pods
+	Paused bool `json:"paused,omitempty"`
+	// Partition describe how many Pods should be upgraded during rollout.
+	Partition intstr.IntOrString `json:"partition,omitempty"`
+	// StableRevision record the user-specified stable revision, and we
+	// will scale down stable revision replicaset in the last.
+	StableRevision string `json:"stableRevision,omitempty"`
+}
+
+// DeploymentExtraStatus is extra status field for Advanced Deployment
+type DeploymentExtraStatus struct {
+	// ObservedGeneration record the generation of deployment this status observed.
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+	// UpdatedReadyReplicas the number of pods that has been updated and ready.
+	UpdatedReadyReplicas int32 `json:"updatedReadyReplicas,omitempty"`
+	// UpdatedReadyReplicas the number of pods that has been updated and available.
+	UpdatedAvailableReplicas int32 `json:"updatedAvailableReplicas"`
+	// DesiredUpdatedReplicas is an absolute number calculated based on Partition.
+	// This field is designed to avoid users to fall into the details of algorithm
+	// for Partition calculation.
+	DesiredUpdatedReplicas int32 `json:"desiredUpdatedReplicas,omitempty"`
+}
+
+func SetDefaultDeploymentStrategy(strategy *DeploymentStrategy) {
```

Comment:
if we reuse MaxSurge and MaxUnavailable in the deployment spec, there is no need to set defaults